pkg/networkserver/grpc_gsns.go

```diff
@@ -189,50 +210,94 @@ outer:
 			"mac_version", match.MACState.LoRaWANVersion,
 		))
 
-		gap := uint32(math.MaxUint32)
-		if fCnt == 0 && match.Session.LastFCntUp == 0 &&
-			(len(match.Device.RecentUplinks) == 0 || match.Device.PendingSession != nil) {
-			gap = 0
-		} else if !resetsFCnt(match.Device, ns.defaultMACSettings) {
-			if fCnt == match.Session.LastFCntUp {
-				// TODO: Handle accordingly (https://github.com/TheThingsNetwork/lorawan-stack/issues/356)
-				logger.Debug("Possible uplink retransmission encountered, skip")
-				continue outer
-			} else if fCnt < match.Session.LastFCntUp {
+		switch {
+		case fCnt == match.Session.LastFCntUp:
+			trans := transmissionNumber(up, match.Device.RecentUplinks...)
+			logger = logger.WithFields(log.Fields(
+				"transmission", trans,
+				"nb_trans", match.MACState.CurrentParameters.ADRNbTrans,
+			))
+			if trans > match.MACState.CurrentParameters.ADRNbTrans {
```

Comment:
Note that this duration can be relatively short (i.e. in the order of tens of seconds), as end device stacks will (probably) wait until there's channel time available for all transmissions in case of duty-cycled regions

Reply:
What would be a sensible time frame?
According to the 1.1 spec:
> The delay between the retransmissions is at the discretion of the end-device and MAY be different for each end-device.

Technically, the delay could also be minutes, hours or days, right?
What kind of retransmission delays do devices use in practice?

![2019-05-07-16:46:03-screenshot](https://user-images.githubusercontent.com/12877905/57309169-e435d400-70e7-11e9-84cc-3585fcaab376.png)

Reply:
make it +10 s